pkg/spdk/engine.go

```diff
@@ -1229,6 +1304,388 @@ func (e *Engine) validateAndUpdateReplicaNvme(replicaName string, bdev *spdktype
+		return false, bdevRaidUUID, fmt.Errorf("engine %s raid delete failed", e.Name)
+	}
+
+	return suspendFrontend, bdevUUID, nil
+}
+
+func (e *Engine) expandReplicas(replicaClients map[string]*client.SPDKClient, spdkClient *spdkclient.Client, size uint64) (expansionSuccess bool) {
+	var (
+		errorLock sync.Mutex
+		wg        sync.WaitGroup
+	)
+
+	failedReplica := map[string]error{}
+
+	for replicaName, replicaClient := range replicaClients {
+		wg.Add(1)
+
+		go func(replicaName string, replicaClient *client.SPDKClient) {
+			defer wg.Done()
+			defer func() {
+				if r := recover(); r != nil {
+					e.log.WithField("replica", replicaName).Errorf("Panic during replica expansion: %v", r)
+				}
+			}()
+
+			replicaStatus, ok := e.ReplicaStatusMap[replicaName]
+			if !ok {
+				e.log.WithField("replica", replicaName).Warn("Replica not found in status map")
+				return
+			}
+
+			replica, err := replicaClient.ReplicaGet(replicaName)
+			if err != nil {
+				errorLock.Lock()
+				failedReplica[replicaName] = errors.Wrap(err, "Get replica failure")
+				errorLock.Unlock()
+				return
+			}
+
+			if replica.SpecSize == size {
+				return
+			}
+
+			if err := disconnectNVMfBdev(spdkClient, replicaStatus.BdevName); err != nil {
+				errorLock.Lock()
+				failedReplica[replicaName] = err
+				errorLock.Unlock()
+				return
+			}
+
+			if err := replicaClient.ReplicaExpand(replicaName, size); err != nil {
+				errorLock.Lock()
+				failedReplica[replicaName] = err
+				errorLock.Unlock()
+				return
+			}
+
+		}(replicaName, replicaClient)
+	}
+
+	wg.Wait()
+
+	// Some replicas may have returned an error during expansion due to unexpected issues
+	// (e.g. temporary network glitch, internal error, timeout).
+	// To avoid mistakenly marking those as failed, we perform a single follow-up check
+	// to verify if the replica was actually expanded.
+	//
+	// If ReplicaGet shows the desired size, we consider the expansion successful
+	// and remove it from the failed list.
+	for replicaName := range failedReplica {
+		client := replicaClients[replicaName]
+
+		replica, err := client.ReplicaGet(replicaName)
+		if err != nil {
+			e.log.WithError(err).WithField("replica", replicaName).
+				Warn("Could not verify replica expansion after failure")
+			continue
+		}
+
+		if replica.SpecSize == size {
+			e.log.WithField("replica", replicaName).
+				Info("Replica expansion succeeded despite earlier error, removing from failure list")
+
+			errorLock.Lock()
+			delete(failedReplica, replicaName)
+			errorLock.Unlock()
+		}
+	}
+
+	switch {
+	case len(failedReplica) == 0:
+		// all success
+		e.log.Warn("All replicas expand success")
+		return true
+	case len(failedReplica) == len(e.ReplicaStatusMap):
+		// all fail, no based lvols are resized
+		e.log.WithFields(logrus.Fields{
+			"engineName": e.Name,
+			"volumeName": e.VolumeName,
+		}).Warn("All replicas failed to expand; proceeding with RAID recreation using existing size")
+		return false
```

Comment:
Can we aggregate errors and return the error here? As we discussed, if all replicas fail, it's better to crash the engine and let the salvage feature help recovery.